src/rust-crypto/PerSessionKeyBackupDownloader.ts

```diff
@@ -0,0 +1,480 @@
+     * @param megolmSessionId - The megolm session ID that is missing.
+     */
+    public onDecryptionKeyMissingError(roomId: string, megolmSessionId: string): void {
+        // Several messages encrypted with the same session may be decrypted at the same time,
+        // so we need to be resistant and not query several time the same session.
+        if (this.isAlreadyInQueue(roomId, megolmSessionId)) {
+            // There is already a request queued for this session, no need to queue another one.
+            this.logger.trace(`Not checking key backup for session ${megolmSessionId} as it is already queued`);
+            return;
+        }
+
+        if (this.wasRequestedRecently(megolmSessionId)) {
+            // We already tried to download this session recently and it was not in backup, no need to try again.
+            this.logger.trace(
+                `Not checking key backup for session ${megolmSessionId} as it was already requested recently`,
+            );
+            return;
+        }
+
+        // We always add the request to the queue, even if we have a configuration problem (can't access backup).
+        // This is to make sure that if the configuration problem is resolved, we will try to download the key.
+        // This will happen after an initial sync, at this point the backup will not yet be trusted and the decryption
+        // key will not be available, but it will be just after the verification.
+        // We don't need to persist it because currently on refresh the sdk will retry to decrypt the messages in error.
+        this.queuedRequests.push({ roomId, megolmSessionId });
+
+        // Start the download loop if it's not already running.
+        this.downloadKeysLoop();
+    }
+
+    public stop(): void {
+        this.stopped = true;
+        this.backupManager.off(CryptoEvent.KeyBackupStatus, this.onBackupStatusChanged);
+        this.backupManager.off(CryptoEvent.KeyBackupFailed, this.onBackupStatusChanged);
+        this.backupManager.off(CryptoEvent.KeyBackupDecryptionKeyCached, this.onBackupStatusChanged);
+    }
+
+    /**
+     * Called when the backup status changes (CryptoEvents)
+     * This will trigger a check of the backup configuration.
+     */
+    private onBackupStatusChanged = (): void => {
+        // we want to check configuration
+        this.hasConfigurationProblem = false;
+        this.configuration = null;
+        this.getOrCreateBackupConfiguration(true).then((decryptor) => {
+            if (decryptor) {
+                // restart the download loop if it was stopped
+                this.downloadKeysLoop();
+            }
+        });
+    };
+
+    /** Returns true if the megolm session is already queued for download. */
+    private isAlreadyInQueue(roomId: string, megolmSessionId: string): boolean {
+        return this.queuedRequests.some((info) => {
+            return info.roomId == roomId && info.megolmSessionId == megolmSessionId;
+        });
```

Comment:
I'm a bit concerned that this operation could get a bit slow, if the queue gets long (eg if there is a long-term problem with backup). Indeed, I'm also a bit concerned we could end up eating a lot of memory here.

Maybe we should use a different datastructure to make it easier to query, and/or remove entries after a while?

Also happy to leave this for now, and fix it later.